pkg/expr/reader_test.go

```diff
@@ -0,0 +1,143 @@
+package expr
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data/utils/jsoniter"
+	data "github.com/grafana/grafana-plugin-sdk-go/experimental/apis/data/v0alpha1"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReaderReduceMode(t *testing.T) {
+	testData := []struct {
+		name          string
+		bytes         []byte
+		expectedError string
+	}{
+		{
+			name: "no_settings",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
+					"type": "reduce"
+				}
+			`),
+			expectedError: "",
+		},
+		{
+			name: "mode_dropnn",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
```

Comment:
It struck me that these lines are the same across all tests. I'm not a "Do not repeat yourself" fanatic, but I do think tests are easier to understand and maintain, if the concern of the test is clear - ie you only really need to specify the part that is changing. But I come from Ruby, where I feel this was easier to do.

So here we'd only want to specify:
```
settings: nil
``` 
or 
```
settings: {
"mode": "dropNN"
},
```
or something. BUT I had a play, and I couldn't find an obvious/easy way to do this. Interpolating the relevant part into a larger string seemed somewhat unfortunate I guess (increasing indirection) and I wasn't able to write it into the parsed version of the object, post JSON parsing, because the methods were private.

Reply:
yes, there is no simple way to achieve this.
we could go for something like:
```go
 jsonStart :=  ```<the json before settings>```;
jsonEnd :=  ```<the json after settings>```;
.
.
.
json := jsonStart + ```
   "settings": { "mode": "dropNN" }
``` + jsonEnd

but i think it would be harder to read because of the indirection, as you mentioned.